src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
+    public void rangeCheck() {
+        if (size == elementData.length) {
+            grow();
+        }
+    }
+
+    public void indexCheck(int index) {
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private Object[] grow() {
+        int newCapacity = elementData.length * 3 / 2 + 1;
+        elementData = Arrays.copyOf(elementData, newCapacity);
+        return elementData;
+    }
+
+    public void checkIfElementExists(T element) {
+        boolean found = false;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elementData[i] == null)
+                    || (element != null && element.equals(elementData[i]))) {
+                found = true;
+                break;
+            }
+        }
+        if (!found) {
+            throw new NoSuchElementException("Element not found in the array.");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        rangeCheck();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bound");
+        }
+        rangeCheck();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
-
+    public void addAll(List<T> o) {
+        for (int i = 0; i < o.size(); i++) {
+            this.add(o.get(i));
+        }
```

Comment:
The `addAll` method assumes that the input `List` has a `size` method and a `get` method. Ensure that the `List` interface being implemented includes these methods, or handle the possibility that it may not.